libs/Microsoft.MixedReality.WebRTC.Native/include/interop_api.h

```diff
@@ -329,93 +450,135 @@ struct PeerConnectionConfiguration {
 MRS_API mrsResult MRS_CALL
 mrsPeerConnectionCreate(PeerConnectionConfiguration config,
                         mrsPeerConnectionInteropHandle interop_handle,
-                        PeerConnectionHandle* peerHandleOut) noexcept;
+                        mrsPeerConnectionHandle* peerHandleOut) noexcept;
 
+/// Callbacks needed to allow the native implementation to interact with the
+/// interop layer, and in particular to react to events which necessitate
+/// creating a new interop wrapper for a new native instance (whose creation was
+/// not initiated by the interop, so for which the native instance is created
+/// first).
 struct mrsPeerConnectionInteropCallbacks {
+  /// Construct an interop object for an AudioTransceiver instance.
+  mrsAudioTransceiverCreateObjectCallback audio_transceiver_create_object{};
+
+  /// Finish the construction of the interop object of an AudioTransceiver.
+  mrsAudioTransceiverFinishCreateCallback audio_transceiver_finish_create{};
+
+  /// Construct an interop object for a VideoTransceiver instance.
+  mrsVideoTransceiverCreateObjectCallback video_transceiver_create_object{};
+
+  /// Finish the construction of the interop object of a VideoTransceiver.
+  mrsVideoTransceiverFinishCreateCallback video_transceiver_finish_create{};
+
+  /// Construct an interop object for a RemoteAudioTrack instance.
+  mrsRemoteAudioTrackCreateObjectCallback remote_audio_track_create_object{};
+
+  /// Construct an interop object for a RemoteVideooTrack instance.
+  mrsRemoteVideoTrackCreateObjectCallback remote_video_track_create_object{};
+
   /// Construct an interop object for a DataChannel instance.
   mrsDataChannelCreateObjectCallback data_channel_create_object{};
 };
```

Comment:
Why do transceivers have a "FinishCreate" callback and tracks don't? Seems strangely asymmetric.

Reply:
The usual pattern that I try to follow for creating objects owned by the user is:
1. Create a wrapper (_e.g._ C#) object via a reverse P/Invoke callback
2. Create the actual low-level implementation object and assign the handle to the wrapper
3. Assign the handle of the newly created implementation object to the wrapper via a second reverse P/Invoke callback if everything went right

For remote tracks here (and only remote tracks; note there's not `FinishCreate` for local ones), those are objects created internally by the Google code and owned by the implementation, not the user (in C# they are not disposable, unlike local tracks). So they are always associated with an event, like `TrackAdded`, whose reverse P/Invoke callback will take care of assigning the handle (step 3. above), so there is no need for an explicit `FinishCreate` for those.

Local tracks seem to still use the old pattern where essentially the steps 1 => 3 above are moved to the wrapping layer (_e.g._ C#), so there is a single (forward) P/Invoke callback for step 2. instead of two reverse P/Invoke callbacks for steps 1 and 3. I don't know what's best between the two approaches.

Reply:
This kind of falls within the discussion about https://github.com/microsoft/MixedReality-WebRTC/pull/202#pullrequestreview-366325725. IRL it was proposed to add API calls for the wrapper code to add/get user data o native objects explicitly. This would bypass the need for this mechanism.